docs/tsp/typeServerProtocol.ts

```diff
@@ -0,0 +1,383 @@
+        // Note this is per byte and not per character.
+        start: number;
+        // The length of the node (in number of bytes) in the source file.
+        length: number;
+    }
+
+    // Represents a category of a type, such as class, function, variable, etc.
+    export const enum TypeCategory {
+        Class = 0,
+        Function = 1,
+        Overloaded = 2,
+        Module = 3,
+        Variable = 4,
+        Parameter = 5,
+        Property = 6,
+        TypeAlias = 7,
+        TypeVariable = 8,
+        Generic = 9,
+        Unknown = 10,
+        Any = 11,
+        None = 12,
+        Union = 13,
+    }
+
+    // Flags that describe the characteristics of a type.
+    // These flags can be combined using bitwise operations.
+    export const enum TypeFlags {
+        None = 0,
+        Instantiable = 1 << 0, // Indicates if the type can be instantiated.
+        Instance = 1 << 1, // Indicates if the type represents an instance (as opposed to a class or type itself).
+        Callable = 1 << 2, // Indicates if an instance of the type can be called like a function. (It has a `__call__` method).
+        BuiltIn = 1 << 3, // Indicates if the type is a built-in type (like `int`, `str`, etc.).
+        Literal = 1 << 4, // Indicates if the instance is a literal (like `42`, `"hello"`, etc.).
+        Interface = 1 << 5, // Indicates if the type is an interface (a type that defines a set of methods and properties). In Python this would be a Protocol.
+        Generic = 1 << 6, // Indicates if the type is a generic type (a type that can be parameterized with other types).
+    }
+
+    // Flags that describe the characteristics of a function or method.
+    // These flags can be combined using bitwise operations.
+    export const enum FunctionFlags {
+        None = 0,
+        Async = 1 << 0, // Indicates if the function is asynchronous.
+        Generator = 1 << 1, // Indicates if the function is a generator (can yield values).
+        Abstract = 1 << 2, // Indicates if the function is abstract (must be implemented in a subclass).
+        Static = 1 << 3, // Indicates if the function is static (belongs to the class rather than an instance).
+    }
+
+    // Flags that describe the characteristics of a class.
+    // These flags can be combined using bitwise operations.
+    export const enum ClassFlags {
+        None = 0,
+        Enum = 1 << 0, // Indicates if the class is an enum (a special kind of class that defines a set of named values).
+    }
+
+    export interface Type {
+        // Unique identifier for the type definition within the session.
+        handle: string | number;
+
+        // Essential classification of the symbol.
+        category: TypeCategory;
+
+        // Flags describing the type.
+        flags: TypeFlags;
+
+        // Name of the module the type comes from
+        moduleName: string | undefined;
+
+        // Name of the type.
+        name: string;
```

Comment:
I presume this is the "short" type name, not its fully-qualified name.

Reply:
Yes, I'll make this more explicit.